src/main/java/com/google/devtools/build/lib/util/Fingerprint.java

```diff
@@ -64,18 +63,14 @@ public final class Fingerprint {
 
   // Make novel use of a CodedOutputStream, which is good at efficiently serializing data. By
   // flushing at the end of each digest we can continue to use the stream.
-  private final CodedOutputStream codedOut;
-  private final MessageDigest messageDigest;
+  private final HashFunction hashFunction;
+  private Hasher hasher;
+  private CodedOutputStream codedOut;
 
   /** Creates and initializes a new instance. */
   public Fingerprint(DigestHashFunction digestFunction) {
-    messageDigest = digestFunction.cloneOrCreateMessageDigest();
-    // This is a lot of indirection, but CodedOutputStream does a reasonable job of converting
-    // strings to bytes without creating a whole bunch of garbage, which pays off.
-    codedOut =
-        CodedOutputStream.newInstance(
-            new DigestOutputStream(ByteStreams.nullOutputStream(), messageDigest),
-            /*bufferSize=*/ 1024);
+    hashFunction = digestFunction.getHashFunction();
```

Comment:
somebody's going to care about this, because they likely believe that going directly to the digest instead of using messagedigesthashfunction is better.  they'll want to see SHA v SHA for the old and new code--does that look negligible?